state/protocol/badger/state.go

```diff
@@ -164,43 +149,43 @@ func Bootstrap(
 		}
 
 		// 3) initialize the current protocol state height/view pointers
-		err = transaction.WithTx(state.bootstrapStatePointers(root))(tx)
+		err = transaction.WithTx(bootstrapStatePointers(root))(tx)
 		if err != nil {
 			return fmt.Errorf("could not bootstrap height/view pointers: %w", err)
 		}
 
 		// 4) initialize values related to the epoch logic
-		err = state.bootstrapEpoch(root.Epochs(), !config.SkipNetworkAddressValidation)(tx)
+		err = bootstrapEpoch(setups, commits, root.Epochs(), !config.SkipNetworkAddressValidation)(tx)
 		if err != nil {
 			return fmt.Errorf("could not bootstrap epoch values: %w", err)
 		}
 
 		// 5) initialize spork params
-		err = transaction.WithTx(state.bootstrapSporkInfo(root))(tx)
+		err = transaction.WithTx(bootstrapSporkInfo(root))(tx)
 		if err != nil {
 			return fmt.Errorf("could not bootstrap spork info: %w", err)
 		}
 
-		// 6) set metric values
-		err = state.updateEpochMetrics(root)
+		// 6) set metric values, we pass `false` here since this node has empty storage and doesn't know anything about EFM.
+		err = updateEpochMetrics(metrics, root, false)
 		if err != nil {
 			return fmt.Errorf("could not update epoch metrics: %w", err)
 		}
```

Comment:
Currently nodes cannot dynamically bootstrap after EFM is triggered. This makes sense on the surface, because dynamic bootstrapping is intended to be used by a node which has newly staked to join the network and no such nodes can join in EFM.

However, many node operators use dynamic bootstrapping as a general operational tool now, not only to facilitate newly staked nodes joining. Especially with the many-month-long sporks we have now, it makes syncing a node which is behind for any reason a lot faster.

While working though spork-less recovery from EFM, it would be great to also enable dynamic bootstrapping even while in EFM. 

(Merely a comment, no action required for this PR.)

Reply:
+100 Lets add a todo in the code and please create an issue 🙏. The last thing we need during a temporary EFM is nodes dropping offline because node operators tried to use dynamic bootstrapping and their node is not coming online anymore.

Reply:
https://github.com/onflow/flow-go/pull/5113/commits/14b468963fdfd18e7f5ec12145a1531aa66273a1
and issue: https://github.com/onflow/flow-go/issues/5127